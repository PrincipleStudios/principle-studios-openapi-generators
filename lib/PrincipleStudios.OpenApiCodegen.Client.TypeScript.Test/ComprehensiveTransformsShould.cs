using PrincipleStudios.OpenApi.Transformations;
using PrincipleStudios.OpenApi.TypeScript;
using Snapshooter.Xunit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using static PrincipleStudios.OpenApiCodegen.TestUtils.DocumentHelpers;

namespace PrincipleStudios.OpenApiCodegen.Client.TypeScript
{
    using static OptionsHelpers;
    public class ComprehensiveTransformsShould
    {
        [Trait("Category", "Integration")]
        [MemberData(nameof(ValidFileNames))]
        [Theory]
        public void CoverFullFiles(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            var sb = new StringBuilder();
            foreach (var entry in entries)
            {
                sb.AppendLine($"exports[`{entry.Key}`] = `{entry.SourceText}`;");
                sb.AppendLine();
            }
            Snapshot.Match(sb.ToString(), $"{nameof(ComprehensiveTransformsShould)}.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
            Assert.Empty(diagnostic.Errors);
        }

        [Trait("Category", "Integration")]
        [MemberData(nameof(InvalidFileNames))]
        [Theory]
        public void ReportDiagnosticsForMissingReferences(string name)
        {
            var document = GetDocument(name);
            var options = LoadOptions();

            var transformer = document.BuildTypeScriptOperationSourceProvider("", options);
            OpenApiTransformDiagnostic diagnostic = new();

            var entries = transformer.GetSources(diagnostic).ToArray();

            Snapshot.Match(diagnostic.Errors.Select(err => new { Context = err.Context.ToOpenApiPathContextString(), Message = err.Message }).ToArray(), $"Diagnostics.{TypeScriptNaming.ToTitleCaseIdentifier(name, options.ReservedIdentifiers())}");
        }

        public static IEnumerable<object[]> ValidFileNames =>
            from fileIndex in GetValidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

        public static IEnumerable<object[]> InvalidFileNames =>
            from fileIndex in GetInvalidDocumentIndices()
            select new object[] { GetDocumentName(fileIndex) };

    }
}
