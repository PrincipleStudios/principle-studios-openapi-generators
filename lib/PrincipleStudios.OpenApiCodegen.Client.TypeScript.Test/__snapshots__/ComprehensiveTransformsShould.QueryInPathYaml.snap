exports[`operations/testPath.ts`] = `/*
 * Query In Path
 *
 * Testing query strings in paths
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { HttpHeaders, HttpQuery, RequestOpts, ResponseArgs, throwIfNullOrUndefined, encodeURI, RequestConversion, StandardResponse, TransformRequest } from '@principlestudios/openapi-codegen-typescript';

export type RequestParams = {
    limit?: number | null;
};
export type RequestBodies = {
};

export type StructuredResponses = {
    /** No content */
    204: {
        '': undefined;
    };
};
export type Responses = 
    | StandardResponse<204, '', undefined>;

export function testPath(params: RequestParams): RequestOpts;
export function testPath(params: RequestParams): RequestOpts {
    throwIfNullOrUndefined(params, 'params');

    const headers: HttpHeaders = {};

    const query: HttpQuery = { 
    };
    if (params.limit != null) { query['limit'] = params.limit; }

    return {
        path: '/path?param1=test&param2=test',
        method: 'GET',
        headers,
        query,
    };
}

export const responseConversion = (res: ResponseArgs) => ({
        statusCode: [
            204,
        ].indexOf(res.status) === -1 ? 'other' : res.status,
        mimeType: res.getResponseHeader("Content-Type"),
        data: res.response,
        response: res,
    }) as Responses;

export const conversion: RequestConversion<RequestParams, {}, Responses, 'no-body'> = {
    request: testPath,
    response: responseConversion,
};

`;

exports[`operations/index.ts`] = `/*
 * Query In Path
 *
 * Testing query strings in paths
 *
 * OpenAPI spec version: 1.0.0
 * 
 * Generated by: https://principle.tools
 */
import { conversion as testPath } from './testPath';

export default {
    testPath,
} as const;
`;

